corehq/apps/users/audit/change_messages.py

```diff
@@ -1,91 +1,329 @@
-class UserChangeMessage(object):
+from enum import Enum
+from django.utils.translation import ugettext as _
+
+
+class Change(Enum):
+    SET = 'set'
+    RESET = 'reset'
+    REMOVE = 'remove'
+    ADD = 'add'
+
+
+class UserChangeMessageV1(object):
+    """
+    Each change message to follow the structure
+    {
+        "field": {
+            "change": change_details
+        }
+    }
+    field: could be domain, phone_numbers etc
+    change: either 'set' / 'reset' / 'remove' / 'add'
+    change_details: a Boolean or
+                    a dict with details as needed for the change
+                    a list of a dicts as needed for the change
```

Comment:
Hey @esoergel 

Here is my attempt to this https://github.com/dimagi/commcare-hq/pull/30181/commits/7b252c8c8a73c85f5c6c58dd840ed095fca08f03
let me know if this is how you were imagining this.

Couple of things:
1. We can't do `"field": "program"` because we need fields to be top-level so that we can search for logs for a specific property. More details here https://github.com/dimagi/commcare-hq/pull/30058#discussion_r675411523
2. Still using a list of changes because there can be more than one changes for a property at once. There is no use case right now to support lists but i can't that there will never be unless we force it to only be one.

I like this approach for it reduces the code quite a lot and makes things generic.
It does take us back to storing things that message need and not storing logs with details, as we discussed [here](https://docs.google.com/document/d/15WS2OuB8l8UJePuwnZASIerxIrB2Gfnf1RRt24O3n8M/edit?disco=AAAANIegoeE)

Couple of challenges
1. this makes customization look out of place when params can't simply be put in translation.
Like [here](https://github.com/dimagi/commcare-hq/pull/30181/commits/7b252c8c8a73c85f5c6c58dd840ed095fca08f03#diff-4284bb62f43ca4b0545a51692cec107399e2d26d955097a6893752c50efa4609R128) I had to add two separate translations to have "re-activated" / "deactivated" which is still okay.
Though, I still need to compute params for where I have collection of records like [here](https://github.com/dimagi/commcare-hq/pull/30181/commits/7b252c8c8a73c85f5c6c58dd840ed095fca08f03#diff-4284bb62f43ca4b0545a51692cec107399e2d26d955097a6893752c50efa4609R278) unless I store `locations_info`directly as a param like I did with [phone numbers](https://github.com/dimagi/commcare-hq/pull/30181/commits/7b252c8c8a73c85f5c6c58dd840ed095fca08f03#diff-4284bb62f43ca4b0545a51692cec107399e2d26d955097a6893752c50efa4609R141)

There is possibility we would want location/role/groups etc to be links to view for the object like `Primary location <a href="">Location Name</a>`, so the customization is only going to increase.
2. one minor thing was that the slugs need to be named in a way such that they don't conflict with each other, like "clear_group", "clear_location" etc and I am guessing it would be hard to come up with new name for something simple as "clear_location" for a v2 unless we just think of appending "_v2" for it to be "clear_location_v2". The list of MESSAGE is going to grow quite a lot, though if we could group them like MESSAGE[field][slug] that would solve this point. For example: MESSAGE['role']['clear'], MESSAGE['location']['clear'] would help keep naming precise and also shorter collections under fields.

Let me know what you think.

Reply:
> We can't do "field": "program" because we need fields to be top-level so that we can search for logs for a specific property. More details here User history report UX updates #30058 (comment)

Oh interesting - I thought that would've been possible from this example in the [linked doc](https://www.postgresql.org/docs/current/datatype-json.html#JSON-CONTAINMENT):
```
-- The object with a single pair on the right side is contained within the object on the left side:
SELECT '{"product": "PostgreSQL", "version": 9.4, "jsonb": true}'::jsonb @> '{"version": 9.4}'::jsonb;
```
But I'm not very familiar with it, so I'll leave that to your judgement.  I just did it like that in my example because "flat is better than nested".  The way you've done it seems fine to me too.

> It does take us back to storing things that message need and not storing logs with details, as we discussed here

Is the difference just a matter of framing?  Like, you think of the details first and the message second?  I'm not convinced there's any distinction.

> this makes customization look out of place when params can't simply be put in translation

Good point - I hadn't thought about that.  Your phone numbers example makes sense.  Another possibility would be to use full-fledged templating:
```python
from django.template import Template, Context
MESSAGES = {"add_phone_numbers": "Added phone number(s) {{phone_numbers|join:', '}}"}
message = {
    "slug": "add_phone_numbers",
    "params": {"phone_numbers": ['555-5555', '867-5309']},
}
Template(MESSAGES[message['slug']]).render(Context(message['params']))
```

To your last point, I do think `_v2` is fine, as it makes it clear that it supersedes the old implementation.  I'm not averse to splitting changes up by field - whichever ends up simpler.  Relatedly, what about changes that aren't tied to a particular field, (like creations or deletions) how are those handled?

Reply:
Thanks for the response @esoergel . I will definitely consider the full-fledged templating, if that makes things simpler.

>  Relatedly, what about changes that aren't tied to a particular field, (like creations or deletions) how are those handled?

These messages are mainly for "updates". When a user is created or deleted, we just save the complete user record at that point, because as such there is no "update". And with users, nothing is created as a part of the update, role/locations etc are all already present so there are just "set" on the user.
Does that answer your question or you were talking about any other create/delete that isn't bound to a field?

Reply:
I see, that makes sense.  I was just wondering how we'd deal with updates that aren't tied to a particular field, but if I can't think of any such updates, maybe they're not a real issue.  At any rate, we could just use a constant string that doesn't correspond to a field if such a concern arises, so I'm not concerned.